### PR TITLE
Add changeset support for gitlab

### PR DESCRIPTION
**Pull request by jgoerz**
* Super naive approach, maybe even stupid... just trying to get the ball rolling and looking for feedback/help/advice.
* I'll be out of pocket for 2 weeks, so I won't get to this right away.
* related to issue [#1234](https://github.com/zendesk/samson/issues/1234)

/cc @zendesk/samson

### Tasks
 - [ ] :+1: from team

### References
 - Jira link:

### Risks
- Level: Low/Med/High

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -32,6 +32,7 @@ gem 'omniauth-ldap'
 gem 'omniauth-gitlab'
 gem 'omniauth-bitbucket'
 gem 'octokit'
+gem 'gitlab'
```
Comment (grosser) [inline Gemfile]: should go into the plugins gemspec

File: config/application.rb [hunk 2]
```diff
@@ -121,8 +121,11 @@ class Application < Rails::Application
     config.samson.ldap.bind_dn = ENV["LDAP_BINDDN"].presence
     config.samson.ldap.password = ENV["LDAP_PASSWORD"].presence
 
+    # Configuration for Gitlab
     config.samson.gitlab = ActiveSupport::OrderedOptions.new
     config.samson.gitlab.web_url = deprecated_url.call("GITLAB_URL") || 'https://gitlab.com'
+    config.samson.gitlab.api_url = deprecated_url.call("GITLAB_API_URL") || 'https://gitlab.com/api/v4'
```
Comment (grosser) [inline config/application.rb]: don't need deprecated_url for new urls

Comment (grosser) [reply config/application.rb]: the api url can just come from the web_url + api/v4 ? (inside the code ... no need to configure it afaik)

File: config/initializers/gitlab.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
```
Comment (grosser) [inline config/initializers/gitlab.rb]: should all go into a plugin

File: config/initializers/gitlab.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#   private
+#
+#   def on_complete(response)
+#     if [301, 302].include?(response[:status].to_i)
+#       raise Octokit::RepositoryUnavailable, response
+#     end
+#   end
+# end
+
+token = ENV['GITLAB_TOKEN']
```
Comment (grosser) [inline config/initializers/gitlab.rb]: document that in `.env.example`

File: config/initializers/gitlab.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+#
+# TODO Need to understand this
+#
+# Octokit.middleware = Faraday::RackBuilder.new do |builder|
+#   builder.use Faraday::HttpCache, shared_cache: false, store: Rails.cache, serializer: Marshal
+#   builder.response :logger, Rails.logger
+#   builder.use Octokit::Response::RaiseError
+#   builder.use Octokit::RedirectAsError
+#   builder.adapter Faraday.default_adapter
+# end
```
Comment (grosser) [inline config/initializers/gitlab.rb]: this is adding caching/logging/error handling to octokit ... might be alright to start without ... but ideally leave the TODO

File: config/initializers/gitlab.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+#
+# Octokit.web_endpoint = Rails.application.config.samson.github.web_url
+
+# Log gitlab request timing so it is more obvious what we spent our time on
+Sawyer::Response.prepend(Module.new do
+  def initialize(*)
+    super
+    Rails.logger.info("GITLAB #{@env.method.upcase} (#{timing}s) #{@env.url}")
+  end
+end)
```
Comment (grosser) [inline config/initializers/gitlab.rb]: ideally something consistent with what github does

**Comment by jgoerz** [comment]
Closing this, work should be on #2940

**Pull request closed without merging**